### PR TITLE
add meta description

### PR DESCRIPTION
**Pull request by xmugenx**
**Fixes #0000**

**Changes proposed in this pull request:**
add setDescription to Application.js for meta description.

**Confirmed**

- [ x ] Frontend changes: tested on a local Flarum installation.
- [ x ] Backend changes: tests are green (run `composer test`).

**Required changes:**

- [ x ] Related core extension PRs: flarum/tags

### CODE REVIEW

**Comment by franzliedke** [comment]
This is probably related to #189.

@xmugenx Thanks for your PR! If you create a pull request that does not reference specific issues, please be so kind to provide *why* you are proposing this change. This will help us understand your PR without looking at the code, and therefore process it faster. :smiley: 

Our pull request and issue templates are very detailed for this very reason.

File: js/src/common/Application.js [hunk 1]
```diff
@@ -13,6 +13,7 @@ import mapRoutes from './utils/mapRoutes';
 import RequestError from './utils/RequestError';
 import ScrollListener from './utils/ScrollListener';
 import { extend } from './extend';
+import {getPlainContent, truncate} from "./utils/string";
```
Comment (franzliedke) [inline js/src/common/Application.js]: Please format this according to the conventions in the surrounding lines.

File: js/src/forum/components/DiscussionPage.js [hunk 2]
```diff
@@ -120,6 +120,10 @@ export default class DiscussionPage extends Page {
 
     if (this.discussion) {
       app.setTitle(this.discussion.title());
+      const firstPost = this.discussion.firstPost();
+      if (firstPost) {
+        app.setDescription(firstPost.contentPlain());
```
Comment (franzliedke) [inline js/src/forum/components/DiscussionPage.js]: Isn't this going to be very long for most posts?

File: js/src/common/Application.js [hunk 3]
```diff
@@ -225,6 +227,25 @@ export default class Application {
...
+    if (description !== null && description.length > 0) {
+      description = truncate(getPlainContent(description), 160, 0);
+    }
+    this.description = description;
+    this.updateDescription();
+  }
+
+  updateDescription() {
+    document.head.querySelector('meta[name=description]').content =
+      (this.description ? this.description : this.forum.attribute('description'));
```
Comment (franzliedke) [inline js/src/common/Application.js]: This should probably use the same truncation logic as the backend side.

**Comment by franzliedke** [comment]
And a general question: what is the point of changing the meta tag on the client side? Does this have any effect at all for search engines / other clients reading the meta tag?

File: js/src/forum/components/DiscussionPage.js (hunk 2 above)
Comment (luceos) [reply js/src/forum/components/DiscussionPage.js]: I've searched our codebase `js` but there's no `setDescription` method anywhere, where exactly is this supposed to have an effect?

Comment (luceos) [reply js/src/forum/components/DiscussionPage.js]: @franzliedke isn't setDescription truncating anyway?

Comment (franzliedke) [reply js/src/forum/components/DiscussionPage.js]: Oh, yes. :see_no_evil:

**Comment by xmugenx** [comment]
I just did some reading about SEO for Single Page Applications from google, I think JS part is unnecessary since flarum already render some HTML for search engine to crawl

**Comment by franzliedke** [comment]
@xmugenx In that case, please remove the frontend changes.

Also, may I ask you to encapsulate the generation of the description in a method on the `Discussion` model? That gives us a place to build in extensibility for this feature in the future (e.g. for SEO extensions).

**Comment by askvortsov1** [comment]
Is this necessary considering https://github.com/flarum/core/blob/d492579638fb52dafbfe65f1f36a95eb6047f7f3/src/Frontend/Content/Meta.php?

**Comment by askvortsov1** [comment]
This is superceded by https://github.com/flarum/core/pull/2049, and can probably be closed.

**Comment by franzliedke** [comment]
Indeed. Thanks for getting this started, @xmugenx !

**Pull request closed without merging**